### PR TITLE
task/WG-74: Create Map React Modal

### PR DESCRIPTION
**Pull request by tjgrafft**
## Overview: ##
Create a new map modal in react.  See the current angular "Create new map" modal for comparison.

skip doing the file/project browser as it requires to much work.

Add POST request to Geoapi's /projectsendpoint by adding/using a new hook in react/src/hooks (and due to being blocked by WG-208,  it defaults to always creating a map linked with the users My Data.)

## PR Status: ##

* [x] Ready.
* [ ] Work in Progress.
* [ ] Hold.

## Related Jira tickets: ##

* [WG-74](https://tacc-main.atlassian.net/browse/WG-74)

## Summary of Changes: ##
- Created a new UI for creating a map project (custom modal)
- Added form validation with UI/UX for user (see screenshots)
- Added working POST request that creates new projects (currently using v2 endpoint, but Nathan and I are working on WG-212 to improve backend and get it ready to use v3) 
- Added unit tests for new CreateMapModal component
- system_path is now dynamic based off the currently logged in and authenticated user. This is used on the new modal for Save Location. When we add a File Browser System, this will need to change to include selected files.
- Used tacc's core-styles to update new modal's style and layout. In WG-246 we will update this to use FieldWrapperFormik from Tup's core-wrappers lib. 
- Added a useCreateProject hook to handle the POST request with the form data. 
## Testing Steps: ## 
1. Start up React frontend (npm ci and npm run dev)
2. Go to localhost:4200 in browser
3. On the Main Menu click the new "Create Map" button
4. Once Modal is open click the Submit button (without any input in the fields to verify form validation is working *should see red text prompting user to fill in missing input*)
5. Fill in the three fields (Name, Description, Custom File Name) with test input
6. Try using symbols other than a hyphen and/or underscore in the Custom File Name field and make sure form validation prompts user with correct message (see screenshot)
7. Once all three fields are populated, click the "Create" button. 
8. You should be taken to the map page (this map is just a placeholder from a set fixture, for now but the project uuid in the URL is correct)
9. Open dev tools in Chrome and check the console to see the data that was sent in POST request (if you decided to check the 'Sync Folder' box, you should see both observable and watch_content as true in the payload)
10. Go to Hazmapper prod and check that new project is present in project listing
11. You should also see that your "You have ___ projects" on the Main Menu for react has increased in number by one.

## UI Photos:
Screenshot 2024-01-12 at 11 11 47 AM
Screenshot 2024-01-12 at 11 12 06 AM
Screenshot 2024-01-12 at 11 12 18 AM
Screenshot 2024-01-12 at 11 12 43 AM

## Core Styles Added:
Screenshot 2024-01-25 at 9 40 46 AM

## Notes: ##
This new Modal will require follow-up work relating to [WG-246](https://tacc-main.atlassian.net/browse/WG-246) (using FieldWrapperFormik), as well as adding the feature/behavior where the filename is updated by the map name (unless the user manually changes the filename after). See https://tacc-main.atlassian.net/browse/WG-249.

### CODE REVIEW

File: react/package-lock.json [hunk 1]
```diff
@@ -1,1759 +1,1031 @@
 {
     "name": "hazmapper",
     "version": "0.0.0",
-    "lockfileVersion": 3,
```
Comment (nathanfranklin) [inline react/package-lock.json:4]: can you recreate your package-lock.json.  you should switch to a newer version of npm (via nvm). like npm version 20.x

Comment (tjgrafft) [reply react/package-lock.json:4]: Sure

File: react/src/pages/MainMenu/MainMenu.tsx [hunk 2]
```diff
@@ -1,13 +1,36 @@
...
+  const { mutate: createProject, isLoading: isCreatingProject } =
+    useCreateProject();
+  const [isModalOpen, setIsModalOpen] = useState(false);
+
+  const toggleModal = () => setIsModalOpen(!isModalOpen);
+
+  const handleCreateProject = (projectData) => {
+    createProject(projectData, {
+      onSuccess: () => {
+        // TO-ADD: Handle success (e.g., show success message, refetch projects, etc.)
```
Comment (nathanfranklin) [inline react/src/pages/MainMenu/MainMenu.tsx:23]: we want to navigate to the new map as now we know the new uuid:

```
import { useNavigate } from 'react-router-dom';
import { Project } from '../../types';

....

      onSuccess: (newProject: Project) => {
        navigate(`/project/${newProject.uuid}`);
      },
```

this /project/uuid route is already supported in `react/src/AppRouter.tsx` and bring us to a MapProject page (with fixtures for map data)

**Review by nathanfranklin — CHANGES_REQUESTED** [review]
Looks great.  👍   I left a variety of comments and suggestions.  We can address some in this PR and then create follow on tickets for other ones (i.e. things that take longer)

File: react/src/requests.ts [hunk 3]
```diff
@@ -85,3 +97,26 @@ export function useGet<ResponseType>({
...
+  const client = axios;
+  const state = store.getState();
+  const token = state.auth.token?.token;
+
+  const postUtil = async (requestData: RequestType) => {
+    const response = await client.post<ResponseType>(
+      `${baseUrl}${endpoint}`,
+      requestData,
+      {
+        headers: { Authorization: `Bearer ${token}` },
```
[the remark this answers is not in the document]
Comment (nathanfranklin) [reply react/src/requests.ts]: cool. saw you fixed this already 💯

File: react/package.json [hunk 4]
```diff
@@ -24,10 +24,11 @@
         "@babel/preset-typescript": "^7.21.0",
         "@changey/react-leaflet-markercluster": "^4.0.0-rc1",
         "@reduxjs/toolkit": "^1.8.4",
-        "@tacc/core-styles": "^2.22.6",
+        "@tacc/core-styles": "^2.22.3",
         "@testing-library/react": "^13.4.0",
         "@types/leaflet.markercluster": "^1.5.1",
         "axios": "^1.6.2",
+        "bootstrap": "^5.3.2",
```
[the remark this answers is not in the document]
Comment (tjgrafft) [reply react/package.json:31]: Yes, it should still work with just the cdn. Let me test that though

File: react/src/components/MapModal/MapModal.test.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+describe('MapModal', () => {
+  const dummyOnSubmit = jest.fn();
+
+  const mockUsername = 'mockUser';
+  const mockEmail = 'mockUser@example.com';
+
+  const mockToken = {
+    token: 'mockTokenValue',
+    expires: Date.now() + 1000 * 60 * 60,
+  };
```
[the remark this answers is not in the document]
Comment (tjgrafft) [reply react/src/components/MapModal/MapModal.test.tsx:18]: This part can be deleted once useName (WG-219) becomes available. I wonder if I should just pull that branch into this one, so I can use it for the unit test and MapModal comp?

File: react/src/components/MapModal/MapModal.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                    }}
+                  >
+                    {errors.system_file}
+                  </div>
+                )}
+              </FormGroup>
+              <FormGroup className="row align-items-center">
+                <Label className="col-sm-4">Save Location:</Label>
+                <div className="col-sm-8 text-primary">
+                  {user ? `/${user.name}` : 'Loading...'}
```
[the remark this answers is not in the document]
Comment (tjgrafft) [reply react/src/components/MapModal/MapModal.tsx:149]: Sure, I can add a comment reminding us that this part will change once the File Browser comp is built

File: react/src/pages/MainMenu/MainMenu.tsx [hunk 7]
```diff
@@ -1,13 +1,36 @@
...
   SectionHeader,
+  Button,
 } from '../../core-components';
 import { useProjects } from '../../hooks';
+import MapModal from '../../components/MapModal';
+import useCreateProject from '../../hooks/projects/useCreateProject';
 
 function MainMenu() {
   const { data, isLoading, error } = useProjects();
+  const { mutate: createProject, isLoading: isCreatingProject } =
```
[the remark this answers is not in the document]
Comment (tjgrafft) [reply react/src/pages/MainMenu/MainMenu.tsx]: Done

File: react/src/components/CreateMapModal/CreateMapModal.tsx [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            description: '',
+            system_file: '',
+            system_id: 'designsafe.storage.default',
+            system_path: '',
+            syncFolder: false,
+          }}
+          validationSchema={validationSchema}
+          onSubmit={handleSubmit}
+        >
+          {/*TODO-REACT: Will change to core-style's FormField comp instead of Formik's Form & Field comp*/}
```
Comment (nathanfranklin) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:98]: Can you add the Jira issue here?

and TODO-REACT -> TODO_REACT

File: react/src/components/CreateMapModal/CreateMapModal.tsx [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+});
+
+const CreateMapModal = ({ isOpen, toggle }: CreateMapModalProps) => {
+  const [errorMessage, setErrorMessage] = useState('');
+  const { data: userData } = useAuthenticatedUser();
+  const { mutate: createProject, isLoading: isCreatingProject } =
+    useCreateProject();
+  const navigate = useNavigate();
+
+  const handleCreateProject = (projectData, actions) => {
```
Comment (nathanfranklin) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:43]:
```suggestion
  const handleCreateProject = (projectData: ProjectRequest, actions) => {
```
would be good to use type here.

File: react/src/components/CreateMapModal/CreateMapModal.tsx [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+                  <span className="input-group-text col-sm-3">.hazmapper</span>
+                </div>
+                {/* Alt solution to render error message bc input group was causing text to not display properly */}
+                {touched.system_file && errors.system_file && (
+                  <div className={`${styles['custom-error-message']}`}>
+                    {errors.system_file}
+                  </div>
+                )}
+              </FormGroup>
+              {/* TODO-REACT: This part will change once the FileBrowser component is added*/}
```
Comment (nathanfranklin) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:151]:
```suggestion
              {/* TODO_REACT: This part will change once the FileBrowser component is added. https://tacc-main.atlassian.net/browse/WG-208*/}
```

File: react/src/components/CreateMapModal/CreateMapModal.tsx [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+        system_file: values.system_file,
+        system_id: values.system_id,
+        system_path: `/${userData.username}`,
+      },
+    };
+    handleCreateProject(projectData, actions);
+  };
+  return (
+    <Modal isOpen={isOpen} toggle={toggle}>
+      <ModalHeader toggle={toggle}>Create a New Map</ModalHeader>
```
[the remark this answers is not in the document]
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:84]: This didn't work for me. Don't think ModalHeader has charCode as a prop? I'd have to add it

[the remark this answers is not in the document]
Comment (nathanfranklin) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:84]: hmm. we did it that way in CEP.  wonder what the difference is.  maybe dropped from bootstrap or react-strap?

[the remark this answers is not in the document]
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:84]: I'll look in CEP and see if I can find an example. Normally the close button is built in with the react-strap ModalHeader, but bc we're using a lower version of bootstrap in our project, that btn-close class doesn't exist. But if I can find it in CEP, I'll use that code here

[the remark this answers is not in the document]
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:84]: /* To render modal close button icon as a Cortal icon */
/* CAVEAT: Pass `charCode="&#xe912;"` to `<ModalHeader>` */
.modal-header .close span {
  /* To mimic `.icon` styles without `@extend` or `composes` (unavailable) */
  /* HACK: Copied (and reduced and edited) from `src/styles/trumps/icon...` */
  font-size: 1.5rem; /* bigger to match header text font height (like design) */
  font-family: Cortal-Icons !important;
}

Looks like a custom handling

File: react/src/components/CreateMapModal/CreateMapModal.tsx [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+          );
+        } else if (err?.response?.status === 500) {
+          setErrorMessage('Internal server error. Please contact support.');
+        }
+        actions.setSubmitting(false);
+      },
+    });
+  };
+
+  const handleSubmit = (values, actions) => {
```
Comment (nathanfranklin) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:63]: are the `actions` and `actions.setSubmitting` needed.

File: react/src/types/projects.ts [hunk 13]
```diff
@@ -12,3 +12,11 @@ export interface Project {
   streetview_instances?: any;
 }
 export class Project implements Project {}
+
+export interface ProjectRequest {
+  project: Project;
+  observable?: boolean;
+  watch_content?: boolean;
```
Comment (nathanfranklin) [inline react/src/types/projects.ts:19]: what about making these not optional
```suggestion
  observable: boolean;
  watch_content: boolean;
```

File: react/src/components/CreateMapModal/CreateMapModal.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (nathanfranklin) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:84]: cool. let's ignore and revisit whenever we do another modal.

File: react/src/components/CreateMapModal/CreateMapModal.tsx (hunk 12 above)
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:63]: They're not necessary, but keeping track of the submitting state might be helpful down the road when/if we add the feature to disable the Submit button, while the form is still handling the POST request? But I can take it out and we can add it back at that time?

Comment (nathanfranklin) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:64]: I might be missing something, but I think you have that case covered already by using  `isCreatingProject` and the `Button`'s isLoading parameter which adds a spinner and disables the button:
```
                <Button
                  size="short"
                  type="primary"
                  attr="submit"
                  isLoading={isCreatingProject}
                >
                  Create
                </Button>
```

File: react/src/components/CreateMapModal/CreateMapModal.tsx [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+      onError: (err) => {
+        // Handle error messages while creating new project
+        setErrorMessage('An error occurred while creating the project.');
+        if (err?.response?.status === 409) {
+          setErrorMessage(
+            'That folder is already syncing with a different map.'
+          );
+        } else if (err?.response?.status === 500) {
+          setErrorMessage('Internal server error. Please contact support.');
+        }
```
Comment (nathanfranklin) [inline react/src/components/CreateMapModal/CreateMapModal.tsx:58]: what about just having two types of error messages? one for the 409 and one for just the general error?
```suggestion
        if (err?.response?.status === 409) {
          setErrorMessage(
            'That folder is already syncing with a different map.'
          );
        } else {
          setErrorMessage('An error occurred while creating the project. Please contact support.');
        }
```

File: react/src/components/CreateMapModal/CreateMapModal.tsx (hunk 12 above)
Comment (tjgrafft) [reply react/src/components/CreateMapModal/CreateMapModal.tsx:64]: Oh yep you're right. Forgot we're using the core-style's Button wrapper now. I'll take the code out.

**Review by nathanfranklin — APPROVED** [review]
LGTM 👍

**Pull request merged**